bugsnag-plugin-android-anr/src/main/jni/anr_handler.c

```diff
@@ -1,119 +1,224 @@
+    return NULL;
+  }
+  if (clz_name == NULL) {
+    return NULL;
+  }
+  jclass clz = (*env)->FindClass(env, clz_name);
+  anr_bsg_check_and_clear_exc(env);
+  return clz;
+}
 
-void bsg_notify_anr_detected(JNIEnv *env) {
-  (*env)->CallVoidMethod(env, obj_plugin, mthd_notify_anr_detected);
+jmethodID anr_bsg_safe_get_method_id(JNIEnv *env, jclass clz, const char *name,
+                                     const char *sig) {
+  if (env == NULL || clz == NULL || name == NULL || sig == NULL) {
+    return NULL;
+  }
+  jmethodID methodId = (*env)->GetMethodID(env, clz, name, sig);
+  anr_bsg_check_and_clear_exc(env);
+  return methodId;
 }
 
-bool bsg_configure_anr_jni(JNIEnv *env) {
-  // get a global reference to the AnrPlugin class
+bool configure_anr_jni(
+        JNIEnv *env) { // get a global reference to the AnrPlugin class
   // https://developer.android.com/training/articles/perf-jni#faq:-why-didnt-findclass-find-my-class
+  if (env == NULL) {
+    return false;
+  }
   int result = (*env)->GetJavaVM(env, &bsg_jvm);
   if (result != 0) {
-    // Failed to fetch VM, cannot continue
     return false;
   }
 
-  jclass clz = (*env)->FindClass(env, "com/bugsnag/android/AnrPlugin");
-  mthd_notify_anr_detected = (*env)->GetMethodID(env, clz, "notifyAnrDetected", "()V");
+  jclass clz = anr_bsg_safe_find_class(env, "com/bugsnag/android/AnrPlugin");
+  if (anr_bsg_check_and_clear_exc(env) || clz == NULL) {
+    return false;
+  }
+  mthd_notify_anr_detected = anr_bsg_safe_get_method_id(
+          env, clz, "notifyAnrDetected", "()V");
   return true;
 }
 
-void bsg_handle_sigquit(int signum, siginfo_t *info, void *user_context) {
-  if (enabled) {
-    // invoke a JNI call from the SIGQUIT handler
-    JNIEnv *env;
-    int result = (*bsg_jvm)->GetEnv(bsg_jvm, (void **)&env, JNI_VERSION_1_4);
-
-    if (result == JNI_OK) { // already attached
-      bsg_notify_anr_detected(env);
-    } else if (result == JNI_EDETACHED) { // attach before calling JNI
-      if ((*bsg_jvm)->AttachCurrentThread(bsg_jvm, &env, NULL) == 0) {
-        bsg_notify_anr_detected(env);
-        (*bsg_jvm)->DetachCurrentThread(bsg_jvm); // detach to restore initial condition
+static void notify_anr_detected() {
+  if (!enabled) {
+    return;
+  }
+
+  bool should_detach = false;
+  JNIEnv *env;
+  int result = (*bsg_jvm)->GetEnv(bsg_jvm, (void **)&env, JNI_VERSION_1_4);
+  switch (result) {
+    case JNI_OK:
+      break;
+    case JNI_EDETACHED:
+      result = (*bsg_jvm)->AttachCurrentThread(bsg_jvm, &env, NULL);
+      if (result != 0) {
+        BUGSNAG_LOG("Failed to call JNIEnv->AttachCurrentThread(): %d", result);
+        return;
       }
-    } // All other results are error codes
-  }
-
-  // Invoke previous handler, if a custom handler has been set
-  // This can be conditionally switched off on certain platforms (e.g. Unity) where
-  // this behaviour is not desirable due to Unity's implementation failing with a SIGSEGV
-  if (invokePrevHandler) {
```

Comment:
Does this trigger a SIGSEGV on Unity when an ANR happens?

Reply:
Verified that it doesn't trigger a segfault.

Since we block `SIGQUIT` now, there's no way for any handler Unity installs to get called.